### DISCUSSION TITLE
How to reliably call outgoing APIs to third party services from CRT extensions

### DISCUSSION THREAD

**Discussion by ax365** [microsoft/Dynamics365Commerce.Solutions#152] [Q&A]
I'm developing an extension for CRT/MPOS. 

For each MPOS transaction, there are a range of possible API calls that I can make to an external service. 
For example, for each qualifying transaction, upon adding customer to the cart, I have to call an API to create a new session with the external service.

Due to the high volume of transactions, there is potential for high volume of these API calls. 
I was wondering if the community has some recommendations how to reliably make outgoing API calls from CRT. 
I've read over the following blogs which discourage the use of `HttpClient`: 

- https://www.aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
- https://learn.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests

I'm currently trying to use `IHttpClientFactory` instead. Dependency injection.

Has anyone implemented something similar for API calls? Or have any suggestions on which pattern to use which is optimal and most secure (given the way that CRT extension framework is designed)?

It would be really nice if Microsoft can provide a sample extension for a scenario like this.

![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/138946791/23865aad-8374-429e-9bb0-8d1618a81121)

**Comment by ax365** [comment]
@madyke, @lukedgr is there someone that can help me out with this one? 
I'm wondering how best to use outgoing API calls for CRT extensions.

**Reply by madyke** [reply]
Hi @ax365 - I'm not very familiar with CRT extensions, but @Pikhulya may be able to help.

**Reply by ax365** [reply]
Hi @Pikhulya, are you able to help out with this? Thanks. 
cc: @madyke
